uncertainties/core.py

```diff
@@ -1756,6 +1756,19 @@ def __bool__(self):
 
     ########################################
 
+    def __hash__(self):
+        """
+        Calculates the hash for any AffineScalarFunc object.
+        The hash is calculated from the nominal_value, and the derivatives.
+        
+        Returns:
+            int: The hash of this object
+        """
+
+        items = sorted(self.derivatives.items(), key= lambda x: id(x[0]))
+        ids, values = zip(*map(lambda item: (id(item[0]), item[1]), items))
+        return hash((self._nominal_value, tuple(ids), tuple(values)))
```

Comment:
> In an ideal world, we could just do frozendict(self.derivatives) but there is no frozendict in the standard library.

There is `MappingProxyType`, but I do not know, if this would work.
https://docs.python.org/3/library/types.html#types.MappingProxyType

Reply:
`MappingProxyType` is not hashable, so it will not work.

I think your solution is more elegant, but for some reason, `assert hash((2*a+a)/3)==hash(a)` fails when using your proposal.

Reply:
Did you update `__hash__` for both `AffineScalarFunc` and `Variable`? I suggested updating one above, but both need to be updated consistently. For `Variable` the changed version should be like:

```python
    return hash((self._nominal_value, ((id(self), 1.),)))
```

Reply:
Yes, I did a mistake with the parenthesis.
Now it works :)

Reply:
One thing I realized -- we moved away from `frozenset` because duplicate derivative coefficient values could be combined and lead to more hash collisions, but that was before we switched to have a collection of tuples instead of two collections with ids and coefficients. Instead of

```python
        derivatives = sorted([(id(key), value) for key, value in self.derivatives.items()])
        return hash((self._nominal_value, tuple(derivatives)))
```

we could do

```python
        derivatives = frozenset((id(key), value) for key, value in self.derivatives.items())
        return hash((self._nominal_value, derivatives))
```

since the ids are always unique and so no tuples will get combined in the set. Whether that is more or less readable or efficient, I am not sure. I like not sorting when it's not necessary but that is pretty minor. I was curious if `frozenset.__hash__` used sorting for its implementation, but I checked and [what it does](https://github.com/python/cpython/blob/1d3225ae056245da75e4a443ccafcc8f4f982cf2/Objects/setobject.c#L727) is XOR the hashes of the elements together which is neat because the XOR operation is also order independent.